### PR TITLE
Instance: Add container `finit_module()` syscall interception support

### PR DESCRIPTION
**Pull request by mihalicyn**
This PR introduces support for `finit_module` syscall interception that allows to lazily load Linux kernel modules from inside unprivileged container if module is in the allow list `linux.kernel_modules`

Specification:
https://discourse.ubuntu.com/t/lxd-container-instance-lazy-modules-loading/42837

### CODE REVIEW

File: lxd/instance/instancetype/instance.go [hunk 1]
```diff
@@ -563,9 +563,19 @@ var InstanceConfigKeysContainer = map[string]func(value string) error{
...
 	"linux.kernel_modules": validate.IsAny,
 
+	// lxdmeta:generate(entities=instance; group=miscellaneous; key=linux.kernel_modules.load)
+	// Specify the kernel modules as a comma-separated list.
+	// ---
+	//  type: string
+	//  defaultdesc: `boot`
+	//  liveupdate: no
+	//  condition: container
+	//  shortdesc: Allow to load kernel modules specified in the linux.kernel_modules list using finit_module() syscall interception
```
Comment (ru-fu) [inline lxd/instance/instancetype/instance.go:576]:
```suggestion
	//  shortdesc: How to load kernel modules
```

File: lxd/instance/instancetype/instance.go [hunk 2]
```diff
@@ -563,9 +563,19 @@ var InstanceConfigKeysContainer = map[string]func(value string) error{
 	//  type: string
 	//  liveupdate: yes
 	//  condition: container
-	//  shortdesc: Kernel modules to load before starting the instance
+	//  shortdesc: Kernel modules to load before starting the instance or allow to load if linux.kernel_modules.load=dynamic
 	"linux.kernel_modules": validate.IsAny,
 
+	// lxdmeta:generate(entities=instance; group=miscellaneous; key=linux.kernel_modules.load)
+	// Specify the kernel modules as a comma-separated list.
```
Comment (ru-fu) [inline lxd/instance/instancetype/instance.go:570]:
```suggestion
	// This option specifies how to load the kernel modules that are specified in {config:option}`instance-miscellaneous:linux.kernel_modules`.
	// Possible values are `boot` (load the modules when booting the container) and `dynamic` (intercept the `finit_modules()` syscall and allow a privileged user in the container's user namespace to load the modules).
```

File: lxd/instance/instancetype/instance.go [hunk 3]
```diff
@@ -563,9 +563,19 @@ var InstanceConfigKeysContainer = map[string]func(value string) error{
 	//  type: string
 	//  liveupdate: yes
 	//  condition: container
-	//  shortdesc: Kernel modules to load before starting the instance
+	//  shortdesc: Kernel modules to load before starting the instance or allow to load if linux.kernel_modules.load=dynamic
```
Comment (ru-fu) [inline lxd/instance/instancetype/instance.go:566]:
```suggestion
	//  shortdesc: Kernel modules to load or allow loading
```

File: lxd/instance/instancetype/instance.go [hunk 4]
```diff
@@ -563,9 +563,19 @@ var InstanceConfigKeysContainer = map[string]func(value string) error{
 	//  type: string
```
Comment (ru-fu) [inline lxd/instance/instancetype/instance.go:563]: I can't comment on the lines above, but replace 
```
	// Specify the kernel modules as a comma-separated list.
```
with
```
	// Specify the kernel modules as a comma-separated list.
        //
	// The modules are loaded before the instance starts, or they can be loaded by a privileged user if {config:option}`instance-miscellaneous:linux.kernel_modules.load` is set to `dynamic`.
```

File: doc/api-extensions.md [hunk 5]
```diff
@@ -2371,3 +2371,8 @@ Adds an embedded OpenFGA authorization driver for enforcing fine-grained permiss
 ## `vm_disk_io_limits`
 
 Adds the ability to limit disk I/O for virtual machines.
+
+## `ct_linux_kernel_modules_dynamic_load`
+
+Adds the `linux.kernel_modules.load` container instance configuration option. If `dynamic` value is set then `finit_modules()` syscall will be intercepted and privileged user in the container's user namespace can load Linux kernel modules specified in the
+allow list `linux.kernel_modules`.
```
Comment (ru-fu) [inline doc/api-extensions.md:2378]:
```suggestion
Adds the `linux.kernel_modules.load` container configuration option. If the option is set to `dynamic`, the `finit_modules()` syscall is intercepted and a privileged user in the container's user namespace can load the Linux kernel modules specified in the
allow list `linux.kernel_modules`.
```

File: doc/config_options.txt [hunk 6]
```diff
@@ -1791,9 +1791,21 @@ See {ref}`cluster-evacuate` for more information.
...
+```
+
+```{config:option} linux.kernel_modules.load instance-miscellaneous
+:condition: "container"
+:defaultdesc: "`boot`"
+:liveupdate: "no"
+:shortdesc: "How to load kernel modules"
+:type: "string"
+This option specifies how to load the kernel modules that are specified in {config:option}`instance-miscellaneous:linux.kernel_modules`.
+Possible values are `boot` (load the modules when booting the container) and `dynamic` (intercept the `finit_modules()` syscall and allow a privileged user in the container's user namespace to load the modules).
```
Comment (simondeziel) [inline doc/config_options.txt:1808]: It feels like it is not "how" kernel modules are loaded but more "when" they are.

As such, I would prefer `ondemand` instead of `dynamic` as the later make it sound as if this would happen automatically/if needed. IMHO, `ondemand` better conveys that a privileged user demanded for that module to be loaded.

File: doc/api-extensions.md (hunk 5 above)
Comment (mihalicyn) [reply doc/api-extensions.md]: Fixed, thanks!

File: doc/config_options.txt (hunk 6 above)
Comment (mihalicyn) [reply doc/config_options.txt:1808]: I tend to agree. ;)

cc @tomp

Comment (tomponline) [reply doc/config_options.txt:1808]: yeah that sounds fine to me.

File: doc/api-extensions.md [hunk 7]
```diff
@@ -2388,3 +2388,8 @@ For more information on access control for OIDC clients, see {ref}`fine-grained-
 ## `vm_disk_io_limits`
 
 Adds the ability to limit disk I/O for virtual machines.
+
+## `ct_linux_kernel_modules_dynamic_load`
```
Comment (tomponline) [inline doc/api-extensions.md:2392]: Lets call this `container_syscall_intercept_finit_modules` to align with other intercepted syscalls extensions.

Comment (tomponline) [reply doc/api-extensions.md:2392]: Also please can this commit come first in the PR to aid with backporting - allows me to see what the following commits are about.

File: lxd/main_forksyscallgo.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
+package main
```
Comment (tomponline) [inline lxd/main_forksyscallgo.go:1]: It would be useful to add a comment to this file explaining why it exists compared to the existing forksyscall command.

File: doc/api-extensions.md (hunk 7 above)
Comment (tomponline) [reply doc/api-extensions.md:2392]: Or should it be `finit_module_parse`?

File: lxd/main_forksyscallgo.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	secData, err := sec.Data()
+	if err != nil {
+		return fmt.Errorf("Can't read .modinfo section: %w", err)
+	}
+
+	secNameDataIdx := bytes.Index(secData, []byte("name="))
+	if secNameDataIdx == -1 {
+		return fmt.Errorf(".modinfo section data looks wrong: can't find \"name=\"")
```
Comment (tomponline) [inline lxd/main_forksyscallgo.go:88]: You can use backticks in `fmt.Errorf()` to avoid the escaping of `"` in strings.

E.g.

```
fmt.Errorf(`.modinfo section data looks wrong: can't find "name="`)
```

File: lxd/seccomp/cgo.go [hunk 10]
```diff
@@ -8,4 +8,5 @@ package seccomp
 // #cgo CFLAGS: -Werror=implicit-function-declaration
 // #cgo CFLAGS: -Werror=return-type -Wendif-labels -Werror=overflow
 // #cgo CFLAGS: -Wnested-externs -fexceptions
+// #cgo pkg-config: libcap
```
Comment (tomponline) [inline lxd/seccomp/cgo.go:11]: As these changes are in a separate seccomp package I think it would warrant these changes be in a separate commit explaining what is being changed and why.

File: lxd/seccomp/seccomp.go [hunk 11]
```diff
@@ -1856,6 +1885,188 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+			C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+			return 0
+		}
+
+		return int(-C.EFAULT)
+	}
+
+	paramValues := C.GoString(&cBuf[0])
+
+	// disallow specifying module parameters, it's just not safe
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:1944]: An expanded comment here explaining *why* would be helpful to others.

File: lxd/seccomp/seccomp.go [hunk 12]
```diff
@@ -1856,6 +1885,188 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+	var output bytes.Buffer
+	p := subprocess.NewProcessWithFds(util.GetExecPath(), forksyscallgoargs, nil,
+		&nullWriteCloser{&output}, &nullWriteCloser{&buffer})
+
+	// We want to drop privileges, as we use "debug/elf" Golang
+	// package which is considered as not safe for production use.
+	// By spawning an unprivileged process we can safely use this package.
+	p.SetCreds(s.s.OS.UnprivUID, s.s.OS.UnprivGID)
+
+	// let's set timeout too
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:2032]: Would be good to explain *why* this is here too.

File: doc/api-extensions.md (hunk 7 above)
Comment (simondeziel) [reply doc/api-extensions.md:2392]: While the `finit_module` syscall designation is technically correct, I had never hear of it before. If we want self descriptive API extension names, maybe one of those would be more friendly to non-kernel devs?

* `container_linux_kernel_modules_load_ondemand`
* `container_linux_kernel_modules_ondemand_loading`

In any case, I think we should have either `dynamic` or `ondemand` (whichever we settle on) in the name to make it easier to link with the actual configuration key/value needed.

Comment (tomponline) [reply doc/api-extensions.md:2392]: I think it needs to have interception in it at least so it goes along with the other ones and is clear what is happening.

**Comment by tomponline** [comment]
@mihalicyn this needs to be rebased please

File: lxd/seccomp/cgo.go (hunk 10 above)
Comment (mihalicyn) [reply lxd/seccomp/cgo.go:11]: Have done!

File: lxd/seccomp/seccomp.go (hunk 12 above)
Comment (mihalicyn) [reply lxd/seccomp/seccomp.go]: Done!

File: lxd/seccomp/seccomp.go (hunk 11 above)
Comment (mihalicyn) [reply lxd/seccomp/seccomp.go]: Done!

File: lxd/main_forksyscallgo.go (hunk 9 above)
Comment (mihalicyn) [reply lxd/main_forksyscallgo.go]: Thanks! Fixed!

File: lxd/main_forksyscallgo.go (hunk 8 above)
Comment (mihalicyn) [reply lxd/main_forksyscallgo.go:3]: Have added! Thanks!

File: doc/api-extensions.md (hunk 7 above)
Comment (mihalicyn) [reply doc/api-extensions.md]: Have fixed!

**Review by simondeziel — DISMISSED** [review]

File: lxd/seccomp/seccomp.go [hunk 13]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+		return int(-C.EPERM)
+	}
+
+	if shared.PathExists(fmt.Sprintf("/sys/module/%s", moduleName)) {
+		return int(-C.EEXIST)
+	}
+
+	err = util.LoadModule(moduleName)
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Failed to load module %s", moduleName)
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:2085]: Please can you use %q when including strings in errors (so it quotes the contents), also please can you include the `err` value in the returned error, so something like this: `fmt.Sprintf("Failed to load module %q: %v", moduleName, err)`

File: lxd/seccomp/seccomp.go [hunk 14]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+			module = strings.TrimPrefix(module, " ")
+
+			if module == moduleName {
+				inAllowList = true
+			}
+		}
+	}
+
+	if !inAllowList {
+		ctx["err"] = fmt.Sprintf("Module %s is not in allowlist (linux.kernel_modules)", moduleName)
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:2075]: %q please

File: lxd/seccomp/seccomp.go [hunk 15]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+
+	// Let's set timeout for the helper process.
+	// It's just a precautionary measure to prevent blocakge of a syscall interception
+	// daemon if anything goes wrong inside the debug/elf Golang package internals.
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	err = p.StartWithFiles(timeoutCtx, []*os.File{moduleFile})
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file: %s", err)
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:2047]: Please use %v here so that the errors .Error() function is called if it has one.

File: lxd/seccomp/seccomp.go [hunk 16]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+
+	moduleFileFD, err := unix.Openat(siov.procFd, fmt.Sprintf("fd/%d", fd), unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (unix.Openat): %s", err)
+		return int(-C.EPERM)
+	}
+
+	moduleFile := os.NewFile(uintptr(moduleFileFD), "/proc/<pid>/fd/<fd>")
+	if moduleFile == nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (os.NewFile): %s", err)
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:2015]: Please use %v here

File: lxd/seccomp/seccomp.go [hunk 17]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+
+	// has CAP_SYS_MODULE capability?
+	if !C.lxd_pid_cap_is_set(C.int(targetPID), C.CAP_SYS_MODULE, C.CAP_EFFECTIVE) {
+		ctx["err"] = "requestor process creds lacks CAP_SYS_MODULE"
+		return int(-C.EPERM)
+	}
+
+	moduleFileFD, err := unix.Openat(siov.procFd, fmt.Sprintf("fd/%d", fd), unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (unix.Openat): %s", err)
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:2009]: Please use %v here

File: lxd/seccomp/seccomp.go [hunk 18]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+
+	ctInitUserNS, err := os.Readlink(fmt.Sprintf("/proc/%d/ns/user", containerInitPID))
+	if err != nil {
+		ctx["err"] = "Can't get userns for ct init process"
+		return int(-C.EPERM)
+	}
+
+	reqUserNS, err := os.Readlink(fmt.Sprintf("/proc/%d/ns/user", targetPID))
+	if err != nil {
+		ctx["err"] = "Can't get userns for requestor process"
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:1989]: Please can you include the err in the error message using %v

File: lxd/seccomp/seccomp.go [hunk 19]
```diff
@@ -1875,6 +1885,197 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
...
+
+		return int(-C.EPERM)
+	}
+
+	containerInitPID := int(siov.msg.init_pid)
+	targetPID := int(siov.req.pid)
+
+	ctInitUserNS, err := os.Readlink(fmt.Sprintf("/proc/%d/ns/user", containerInitPID))
+	if err != nil {
+		ctx["err"] = "Can't get userns for ct init process"
```
Comment (tomponline) [inline lxd/seccomp/seccomp.go:1983]: Please can you include the err in the error message using %v

**Comment by tomponline** [comment]
@mihalicyn did you update the spec to reflect the new "ondemand" mode rather than "dynamic"?

**Review by tomponline — CHANGES_REQUESTED** [review]
Looks good thanks! just a few nits to change in error messages.

Also, whats the plan with automated testing of this feature?

Are you thinking adding something to lxd-ci repo?

File: lxd/seccomp/seccomp.go (hunk 15 above)
Comment (mihalicyn) [reply lxd/seccomp/seccomp.go]: Fixed! Thanks!

File: lxd/seccomp/seccomp.go (hunk 14 above)
Comment (mihalicyn) [reply lxd/seccomp/seccomp.go]: Fixed! Thanks!

File: lxd/seccomp/seccomp.go (hunk 13 above)
Comment (mihalicyn) [reply lxd/seccomp/seccomp.go]: Fixed!

**Comment by mihalicyn** [comment]
> @mihalicyn did you update the spec to reflect the new "ondemand" mode rather than "dynamic"?

yes

>Also, whats the plan with automated testing of this feature?

Sure, I'll prepare something for lxd-ci repo once this PR will be merged

**Review by tomponline — APPROVED** [review]
LGTM thanks!

File: doc/config_options.txt [hunk 20]
```diff
@@ -1791,9 +1791,21 @@ See {ref}`cluster-evacuate` for more information.
...
 :liveupdate: "yes"
-:shortdesc: "Kernel modules to load before starting the instance"
+:shortdesc: "Kernel modules to load or allow loading"
 :type: "string"
 Specify the kernel modules as a comma-separated list.
+
+The modules are loaded before the instance starts, or they can be loaded by a privileged user if {config:option}`instance-miscellaneous:linux.kernel_modules.load` is set to `ondemand`.
+```
+
+```{config:option} linux.kernel_modules.load instance-miscellaneous
```
Comment (simondeziel) [inline doc/config_options.txt:1801]: Could you add that extra key to `scripts/bash/lxd-client`, please?

**Review by simondeziel — CHANGES_REQUESTED** [review]
Could you add the new `linux.kernel_modules.load` key to `scripts/bash/lxd-client`, please?

**Comment by mihalicyn** [comment]
> Could you add the new `linux.kernel_modules.load` key to `scripts/bash/lxd-client`, please?

Have done! Thanks!

**Review by simondeziel — APPROVED** [review]
LGTM, thanks!

**Review by tomponline — APPROVED** [review]

**Comment by mihalicyn** [comment]
Test for it
https://github.com/canonical/lxd-ci/pull/131
